### ISSUE TITLE
0.1.47 changed matching of `\r`

### ISSUE THREAD

**Issue by DanielKeep** [rust-lang-nursery/regex#153]
Given the following program:

```rust
extern crate regex;
fn main() {
    let re = regex::Regex::new(r"^(.*?)(\n|\r\n?|$)").unwrap();
    println!("{:?}", re.find("ab\rcd"));
}
```

and rustc nightly 2bd875d3d4a9f06a4a18325d6e4a68b5234a09f6 (2016-01-19), version 0.1.46 of regex outputs:

```
Some((0, 3))
```

Version 0.1.47 outputs:

```
Some((0, 5))
```

I'm not sure what's going on here, but my understanding is that the new interpretation is wrong.

**Comment by DanielKeep** [comment]
Also, I just tested `r"^(.*?)(\n|\r\n|\r|$)"`, and it produces `Some((0, 3))` on both 0.1.46 and 0.1.47.

**Comment by BurntSushi** [comment]
This is indeed a bug! I'm shocked this wasn't caught in the test suite.

I've got a smaller example: `ab?|$` on text `az` should return `(0, 1)` (and does on `0.1.46`), but returns `Some((2, 2))` on `0.1.47`. Specifically, the compilation on `0.1.46` is:

```
000 Save(0)
001 Split(2, 6)
002 'a'
003 Split(4, 5)
004 'b'
005 Jump(7)
006 EndText
007 Save(1)
008 Match
```

And the compilation on `0.1.47` is:

```
000 Save(0)
001 Split(2, 5)
002 'a'
003 Split(4, 5)
004 'b' (goto: 6)
005 EndText
006 Save(1)
007 Match
```

The second compilation is wrong. The `Split(4, 5)` instruction *should* be `Split(4, 6)`. Splitting to `5` means that either a `b` or a `$` is required after matching an `a`, but of course, an `a` is sufficient for a match.

In the case of your regex, this causes the alternation to fail and the preceding `.*?` ends up matching everything.

Your workaround fixed things because it removed the inner alternation which is provoking this bug.

Working on a fix!

**Comment by BurntSushi** [comment]
Fixed in `0.1.48` on crates.io. Thanks for the bug report @DanielKeep!

**Comment by DanielKeep** [comment]
Don't mention it; thanks for regex.  :)
